### PR TITLE
feat: Reconcile on change of argocd-repo-server-tls and implement AutoTLS feature

### PR DESCRIPTION
**Pull request by jannfis**
**What type of PR is this?**

> /kind enhancement

**What does this PR do / why we need it**:

This PR introduces two new configurables:

* `.spec.repo.stricttls: <bool>` : If set to `true`, will configure `<instance>-server` and `<instance>-application-controller` to validate the TLS certificate of the `<instance>-repo-server`'s gRPC endpoint, by adding `--repo-server-strict-tls` to the pods command's arguments.
* `.spec.repo.autotls: <string>`: Can be set to a provider which does automatic provisioning of the `argocd-repo-server-tls` secret. Currently, the only provider supported is `openshift`

The main feature supported is to watch for changes to the `argocd-repo-server-tls` secret, containing the TLS configuration (cert/key combination) for the gRPC API of the repository server. If a change is detected, the appropriate workloads will be restarted. If the `argocd-repo-server-tls` was created using `AutoTLS` feature, no further configuration will be necessary. If the secret was manually added to the namespace, it will require an annotation `argocds.argoproj.io/name: <instance_name>` to be successfully mapped back to the managing `ArgoCD` instance.

Few words about `AutoTLS` feature on OpenShift:

OpenShift Container Platform provides a cluster-internal CA to issue service certificates, as described [here](https://docs.openshift.com/container-platform/4.7/security/certificates/service-serving-certificate.html#understanding-service-serving_service-serving-certificate). If `.spec.repo.autotls: openshift` is specified, the controller will add the appropriate annotation to the `<instance>-repo-server` Service in the instance's namespace, which indicates that a TLS certificate for this service should be issued by the service CA.

**Have you updated the necessary documentation?**

* [ ] Documentation update is required by this PR.
* [ ] Documentation has been updated.

**Which issue(s) this PR fixes**:

Fixes #?

**How to test changes / Special notes to the reviewer**:

To test the `stricttls` feature (can be tested on any platform, plain K8s and OpenShift alike)

* Create ArgoCD CR with all default values:
  ```yaml
   apiVersion: argoproj.io/v1alpha1
   kind: ArgoCD
   metadata:
     name: argocd
  ```
* Wait until reconciliation is complete and workloads became available

* Modify the ArgoCD CR to set `.spec.repo.stricttls` to `true`:
  ```
  kubectl patch argocds.argoproj.io argocd --type=merge -p '{"spec":{"repo":{"stricttls":true}}}'
  ```
* Check `argocd-server` deployment for containing `--repo-server-strict-tls` in the container's command:
  ```
  kubectl get deployments.apps argocd-server -o jsonpath='{.spec.template.spec.containers[0].command}'
  ```
* Check `argocd-application-controller` statefulset for containing `--repo-server-strict-tls` in the container's command:
  ```
  kubectl get statefulsets.apps argocd-application-controller -o jsonpath='{.spec.template.spec.containers[0].command}'
  ```
* Pods for `argocd-server` and `argocd-application-controller` should have been restarted by that change:
  ```
  kubectl get pods
  ```

To test the `autotls` feature (works only on OpenShift):

* Create ArgoCD CR with all default values:
  ```yaml
   apiVersion: argoproj.io/v1alpha1
   kind: ArgoCD
   metadata:
     name: argocd
  ```
* Wait until reconciliation is complete and workloads became available
* Modify the ArgoCD CR to set `.spec.repo.autotls` to `openshift`:
  ```
  kubectl patch argocds.argoproj.io argocd --type=merge -p '{"spec":{"repo":{"autotls":"openshift"}}}'
  ```
* Annotate the `argocd-repo-server` service so that the OpenShift service CA will issue a certificate into `argocd-repo-server-tls`:
  ```
  kubectl annotate service argocd-repo-server service.beta.openshift.io/serving-cert-secret-name=argocd-repo-server-tls
  ```
* Verify the `argocd-repo-server-tls` secret has been created by the CA's controller:
  ```
  kubectl get secrets argocd-repo-server-tls
  ```
  The secret should be of type `kubernetes.io/tls` and should have a size of `2`
* Verify the workloads `argocd-server`, `argocd-repo-server` and `argocd-application-controller` have been restarted
* For each pod of `argocd-server`, `argocd-repo-server` and `argocd-application-controller`, do a `kubectl exec -it <podname> bash` and look for the following files to exist there:
  * argocd-server: `/app/config/server/tls/tls.crt`
  * argocd-repo-server: `/app/config/reposerver/tls/tls.crt`
  * argocd-application-controller: `/app/config/controller/tls/tls.crt`

To test reconciliation of `argocd-repo-server-tls` on non-OpenShift platform, or without `autotls` enabled:

* Create ArgoCD CR with all default values:
  ```yaml
   apiVersion: argoproj.io/v1alpha1
   kind: ArgoCD
   metadata:
     name: argocd
  ```
* Wait until reconciliation is complete and workloads became available
* Create a self-signed certificate for the repository server (crt and key will be stored in `/tmp/reposerver.crt` and `/tmp/reposerver.key` respectively):
  ```
  openssl req -new -x509 -sha256 \
    -subj "/C=XX/ST=XX/O=Testing/CN=repo-server" \
    -reqexts SAN -extensions SAN \
    -config <(printf "\n[SAN]\nsubjectAltName=DNS:argocd-repo-server.$NAMESPACE.svc.cluster.local\n[req]\ndistinguished_name=req") \
    -keyout /tmp/reposerver.key \
    -out /tmp/reposerver.crt \
    -newkey rsa:4096 \
    -nodes \
    -sha256 \
    -days 10
  ```
* Create the `argocd-repo-server-tls` secret of type `kubernetes.io/tls`:
  ```
  kubectl create secret tls argocd-repo-server-tls --key=/tmp/reposerver.key --cert=/tmp/reposerver.crt
  ```
* Check that pods of `argocd-server`, `argocd-repo-server` and `argocd-application-controller` were *not* restarted just yet
* Annotate the `argocd-repo-server-tls` secret to indicate it belongs to the `argocd` operand:
  ```
  kubectl annotate secret argocd-repo-server-tls argocds.argoproj.io=argocd
  ```
* Check that the Operator reconciled and the pods of `argocd-server`, `argocd-repo-server` and `argocd-application-controller` have been restarted.

### CODE REVIEW

**Comment by jannfis** [comment]
This has a dependency to #294

File: docs/reference/argocd.md [hunk 1]
```diff
@@ -733,6 +733,8 @@ Name | Default | Description
 Resources | [Empty] | The container compute resources.
 MountSAToken | false | Whether the ServiceAccount token should be mounted to the repo-server pod.
 ServiceAccount | "" | The name of the ServiceAccount to use with the repo-server pod.
+StrictTLS | false | Whether to enforce strict TLS checking on all components when communicating with repo server
```
Comment (wtam2018) [inline docs/reference/argocd.md:736]: This is really `TLSVerify`, right?  Enable to validation TLS cerificate.

**Comment by wtam2018** [comment]
You need to rebase to pick up https://github.com/argoproj-labs/argocd-operator/pull/313 then run hack/generate.sh (to re-generate zz_generated files).

File: pkg/controller/argocd/deployment.go [hunk 2]
```diff
@@ -1028,13 +1032,13 @@ func (r *ReconcileArgoCD) reconcileServerDeployment(cr *argoprojv1a1.ArgoCD) err
 }
 
 // triggerRollout will update the label with the given key to trigger a new rollout of the Deployment.
```
Comment (wtam2018) [inline pkg/controller/argocd/deployment.go:1034]:
```suggestion
// triggerDeploymentRollout will update the label with the given key to trigger a new rollout of the Deployment.
```

File: pkg/controller/argocd/statefulset.go [hunk 3]
```diff
@@ -462,3 +465,14 @@ func (r *ReconcileArgoCD) reconcileStatefulSets(cr *argoprojv1a1.ArgoCD) error {
 	}
 	return nil
 }
+
+// triggerRollout will update the label with the given key to trigger a new rollout of the Deployment.
```
Comment (wtam2018) [inline pkg/controller/argocd/statefulset.go:469]:
```suggestion
// triggerStatefulSetRollout will update the label with the given key to trigger a new rollout of the StatefulSet.
```

File: pkg/controller/argocd/service.go [hunk 4]
```diff
@@ -307,13 +307,39 @@ func (r *ReconcileArgoCD) reconcileRedisService(cr *argoprojv1a1.ArgoCD) error {
...
+	autoTLSAnnotationName := ""
+	if cr.Spec.Repo.AutoTLS == "openshift" {
+		autoTLSAnnotationName = "service.beta.openshift.io/serving-cert-secret-name"
+	}
+	if autoTLSAnnotationName != "" {
+		if svc.Annotations == nil {
+			svc.Annotations = make(map[string]string)
+		}
+		val, ok := svc.Annotations[autoTLSAnnotationName]
+		if !ok || val != "argocd-repo-server-tls" {
```
Comment (wtam2018) [inline pkg/controller/argocd/service.go:320]: Hmm ...  we have a const ArgoCDRepoServerTLSSecretName.   Why not using it?

File: docs/reference/argocd.md (hunk 1 above)
Comment (jannfis) [reply docs/reference/argocd.md:736]: Yes, I do agree with you, @wtam2018. Will change it to `TLSVerify`.

File: pkg/controller/argocd/service.go (hunk 4 above)
Comment (jannfis) [reply pkg/controller/argocd/service.go:320]: Good catch, thank you. Will update.

Comment (jannfis) [reply pkg/controller/argocd/service.go]: Updated.

File: pkg/apis/argoproj/v1alpha1/zz_generated.openapi.go [hunk 5]
```diff
@@ -27,12 +27,12 @@ import (
```
Comment (jannfis) [inline pkg/apis/argoproj/v1alpha1/zz_generated.openapi.go:27]: This is a revert of a change introduced in #313 which no one is sure where it came from.

File: docs/reference/argocd.md (hunk 1 above)
Comment (jannfis) [reply docs/reference/argocd.md]: Changed it to `VerifyTLS` instead.

**Comment by jannfis** [comment]
Thanks for the review, @wtam2018!

> You need to rebase to pick up #313 then run hack/generate.sh (to re-generate zz_generated files).

Done that and addressed all other review comments. While rebasing to #313, I found that it has introduced a weird change in generated package names that nobody can explain so far. While running openapi-gen here, it reverted the change back to its original package names. I submitted these change as well as part of this PR.

PTAL @wtam2018

File: docs/reference/argocd.md [hunk 6]
```diff
@@ -750,6 +752,8 @@ spec:
     resources: {}
     mountsatoken: false
     serviceaccount: ""
+    stricttls: false
```
Comment (wtam2018) [inline docs/reference/argocd.md:755]:
```suggestion
    verifytls: false
```

**Review by wtam2018 — APPROVED** [review]
LGTM
thanks @jannfis

**Pull request merged**